README.md
## jira-node-client - A node client to easily get data from Jira

[![Build Status](https://secure.travis-ci.org/clandriot/jira-node-client.png?branch=master)](http://travis-ci.org/clandriot/jira-node-client) [![Coverage Status](https://coveralls.io/repos/github/clandriot/jira-node-client/badge.svg)](https://coveralls.io/github/clandriot/jira-node-client) [![Dependency Status](https://www.versioneye.com/user/projects/587cc73f5a9a49003f91d136/badge.svg?style=flat-square)](https://www.versioneye.com/user/projects/587cc73f5a9a49003f91d136) [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/clandriot/jira-node-client/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/clandriot/jira-node-client/?branch=master)

jira-node-client is a node module providing a function to easily get data from Jira with pages aggreation and retries features.

#### Features
* executes provided Jira rest api and return response
* if response includes several pages (total > maxResults) all pages optionally retrieved and aggregated in the response
* In case of specific errors (5xx, ECONNRESET, ENOTCONN or if the body of the error is 'timeout exceeded'), several attempts are done with a timeout between each ones (5 attempts, 1 second timeout)

#### Installation
```
npm install jira-node-client
```

#### How to use?
First, make sure credentials are set as environment variables:
* JIRA_USER
* JIRA_PASSWORD

##### To check if credentials are set
```javascript
const jira = require('jira-node-client');
jira.areJiraCredentialsMissing(); // returns true if any of the 2 credentials is not set
```

#### Authentication
Default behavior is basic authentication, meaning that user and password is set in the header of each request to Jira.
But you can also enable cookie based authentication, meaning that a first login will retrieve a cookie that will be added to all further api calls. If you application runs long enough to reach cookie expiration, this is detected and a new cookie is automatically retrieved. This will so remain transparent.
To enable cookie based authentication
```javascript
const jira = require('jira-node-client');
jira.config.authentication = 'cookie'; // enable cookie based authentication
```

##### To get data
```javascript
const jira = require('jira-node-client');
jira.execJiraQuery('https://www.myjirainstance.com/jira/rest/api/latest/search?jql=project = PROJKEY and issuetype not in (Epic,subTaskIssueTypes()) and resolution != Unresolved', true)
  .then((response) => { // response is the JSON object returned by the api
    let nbIssues = response.total;
    let firstIssueKey = response.issues[0].key;
  })
  .catch((error) => {
    // error is an object including 2 properties: code and message
  });
```
Note: Jira REST Apis and JSON format can be found at https://docs.atlassian.com/jira/REST/cloud/.

##### To configure
You can configure retry behavior
```javascript
const jira = require('jira-node-client');
jira.config.maxRetry = 10 // On server errors, retries to execute the api 10 time; default is 5
jira.config.retryTimeout = 5000 // wait 5000ms between each retry; default is 1000ms
```

#### Running tests
To run the tests suite, first install devlopment dependencies:
```
npm install
```
Then run the tests:
```
npm test
```
To check code coverage, run:
```
npm run cover
```

#### To do
- [x] provide function to check that credentials are set in environment
- [ ] provide cookie based authentication
- [ ] base url is configured once
- [x] make timeout and number of retries configurable


CHANGELOG.md
# 1.1.0 - 2017-01-26
- Support for cookie based authentication

# 1.0.1 - 2017-01-16
- No code change, only build management (see badges)

# 1.O.O - 2017-01-11
- add configuration options to set number of retries and timeout


test/test-jira.js
const test = require('unit.js');
const sinon = require('sinon');
const chai = require('chai');
const expect = chai.expect;
const chaiAsPromised = require('chai-as-promised');
const rewire = require('rewire');
const requestify = require('requestify');
const jira = rewire('../lib/jira');
const cookieAuth = require('../lib/cookieAuthentication');
const testThat = test.promise;
let sandbox = null;

before(() => {
  chai.use(chaiAsPromised);
  jira.config.retryTimeout = 10;
  jira.config.authentication = 'basic';
});

beforeEach(() => {
  sandbox = sinon.sandbox.create();
  jira.__set__('sessionCookie', null);
});

afterEach(() => sandbox.restore());

describe('buildUrlToGetNextPage', () => {
  let buildUrl = jira.__get__('buildUrlToGetNextPage');
  it('buildUrlToGetNextPage() appends &startAt=50 to base url', () => testThat
      .given(() => 'http://host/any')
      .when((oldUrl) => buildUrl(oldUrl, 50))
      .then((newUrl) => expect(newUrl).to.be.equal('http://host/any&startAt=50'))
  );
  it('buildUrlToGetNextPage() replace existing startAt with new one', () => testThat
      .given(() => 'http://host/any&startAt=100')
      .when((oldUrl) => buildUrl(oldUrl, 150))
      .then((newUrl) => expect(newUrl).to.equal('http://host/any&startAt=150'))
  );
});

describe('areJiraCredentialsMissing', () => {
  it('areJiraCredentialsMissing() returns true without JIRA_USER and JIRA_PASSWORD', () => {
    delete process.env.JIRA_USER;
    delete process.env.JIRA_PASSWORD;
    testThat
      .given()
      .when(() =>jira.areJiraCredentialsMissing())
      .then((status) => expect(status).to.be.true);
  });
  it('areJiraCredentialsMissing() returns true with only JIRA_PASSWORD missing', () => {
    process.env.JIRA_USER = 'user';
    delete process.env.JIRA_PASSWORD;
    testThat
      .given()
      .when(() =>jira.areJiraCredentialsMissing())
      .then((status) => expect(status).to.be.true);
  });
  it('areJiraCredentialsMissing() returns true with only JIRA_USER missing', () => {
    process.env.JIRA_PASSWORD = 'pwd';
    delete process.env.JIRA_USER;
    testThat
      .given()
      .when(() =>jira.areJiraCredentialsMissing())
      .then((status) => expect(status).to.be.true);
  });
  it('areJiraCredentialsMissing() returns false if JIRA_USER and JIRA_PASSWORD are set', () => {
    process.env.JIRA_PASSWORD = 'pwd';
    process.env.JIRA_USER = 'user';
    testThat
      .given()
      .when(() =>jira.areJiraCredentialsMissing())
      .then((status) => expect(status).to.be.false);
  });
});

describe('execJiraQuery', () => {
  it('execJiraQuery returns parsed response on 200', () => {
    sandbox.stub(requestify, 'get', () => {
      return Promise.resolve({code: 200, body: '{"message": "successful!!"}'});
    });
    return expect(jira.execJiraQuery('', false)).to.eventually.have.property('message', 'successful!!');
  });
  it('execJiraQuery returns error on 400', () => {
    sandbox.stub(requestify, 'get', () => {
      return Promise.reject({code: 400, body: 'Failed!'});
    });
    return expect(jira.execJiraQuery('', false)).to.eventually.be.rejected;
  });
  it('execJiraQuery returns error on 500', function() {
    // this.timeout(8000); // need to increase Mocha timeout to give enough time to the retries
    sandbox.stub(requestify, 'get', () => {
      return Promise.reject({code: 500, body: 'Failed!'});
    });
    return expect(jira.execJiraQuery('', false)).to.eventually.be.rejected;
  });
  it('execJiraQuery returns error on 502', function() {
    // this.timeout(8000); // need to increase Mocha timeout to give enough time to the retries
    sandbox.stub(requestify, 'get', () => {
      return Promise.reject({code: 502, body: 'Failed!'});
    });
    return expect(jira.execJiraQuery('', false)).to.eventually.be.rejected;
  });
  it('execJiraQuery fails with a not JSON response', () => {
    sandbox.stub(requestify, 'get', () => {
      return Promise.resolve({code: 200, body: 'Not JSON'});
    });
    return expect(jira.execJiraQuery('', false)).to.eventually.be.rejected
      .then((error) => {
        expect(error).to.have.property('code', -1);
        expect(error).to.have.property('message', 'Unexpected token N in JSON at position 0');
      });
  });
  it('execJiraQuery aggregates results  if retrieveAllPages = true and if response specifies multiple pages', () => {
    let first = '{"resultset": "first"}';
    let second = '{"resultset": "second"}';
    let last = '{"resultset": "last"}';
    sandbox.stub(requestify, 'get', (url) => {
      if (url === '') {
        return Promise.resolve({code: 200, body: '{"startAt": 0,"total": 12,"maxResults": 5,"issues": [' + first + ']}'});
      } else if (url === '&startAt=5') {
        return Promise.resolve({code: 200, body: '{"startAt": 5,"total": 12,"maxResults": 5,"issues": [' + second + ']}'});
      } else {
        return Promise.resolve({code: 200, body: '{"startAt": 10,"total": 12,"maxResults": 5,"issues": [' + last + ']}'});
      }
    });
    return expect(jira.execJiraQuery('', true)).to.eventually.be.fulfilled
      .then((response) => {
        expect(response).to.have.property('startAt', 0);
        expect(response).to.have.property('total', 12);
        expect(response).to.have.property('maxResults', 12);
        expect(response).to.have.property('issues');
        expect(response.issues).to.be.an('array');
        expect(response.issues).to.have.length(3);
        expect(response.issues[0]).to.have.property('resultset', 'first');
        expect(response.issues[1]).to.have.property('resultset', 'second');
        expect(response.issues[2]).to.have.property('resultset', 'last');
      });
  });
  it('execJiraQuery doesn\'t aggregate results if retrieveAllPages = false', () => {
    let first = '{"resultset": "first"}';
    let second = '{"resultset": "second"}';
    let last = '{"resultset": "last"}';
    sandbox.stub(requestify, 'get', (url) => {
      if (url === '') {
        return Promise.resolve({code: 200, body: '{"startAt": 0,"total": 12,"maxResults": 5,"issues": [' + first + ']}'});
      } else if (url === '&startAt=5') {
        return Promise.resolve({code: 200, body: '{"startAt": 5,"total": 12,"maxResults": 5,"issues": [' + second + ']}'});
      } else {
        return Promise.resolve({code: 200, body: '{"startAt": 10,"total": 12,"maxResults": 5,"issues": [' + last + ']}'});
      }
    });
    return expect(jira.execJiraQuery('', false)).to.eventually.be.fulfilled
      .then((response) => {
        expect(response).to.have.property('startAt', 0);
        expect(response).to.have.property('total', 12);
        expect(response).to.have.property('maxResults', 5);
        expect(response).to.have.property('issues');
        expect(response.issues).to.be.an('array');
        expect(response.issues).to.have.length(1);
        expect(response.issues[0]).to.have.property('resultset', 'first');
      });
  });
  it('execJiraQuery is rejected if failure occus during aggregation', () => {
    let first = '{"resultset": "first"}';
    sandbox.stub(requestify, 'get', (url) => {
      if (url === '') {
        return Promise.resolve({code: 200, body: '{"startAt": 0,"total": 8,"maxResults": 5,"issues": [' + first + ']}'});
      } else if (url === '&startAt=5') {
        return Promise.reject({code: 400, body: 'Failed!'});
      }
    });
    return expect(jira.execJiraQuery('', true)).to.eventually.be.rejected;
  });

  describe('execJiraQuery with CBA', () => {
    it('execJiraQuery returns parsed response on 200 with CBA', () => {
      jira.config.authentication = 'cookie';
      sandbox.stub(cookieAuth, 'login', () => {
        let header = {'set-cookie': ['mycookie=RUTN87766HG']};
        let body = '{"session": {"name": "mycookie"}}';
        return Promise.resolve({code: 200, headers: header, body: body});
      });
      sandbox.stub(requestify, 'get', () => {
        return Promise.resolve({code: 200, body: '{"message": "successful!!"}'});
      });
      return expect(jira.execJiraQuery('', true)).to.eventually.be.fulfilled;
    });
    it('execJiraQuery rejects if can\'t get cookie', () => {
      jira.config.authentication = 'cookie';
      sandbox.stub(cookieAuth, 'login', () => {
        return Promise.reject(new Error('can\'t get cookie'));
      });
      return expect(jira.execJiraQuery('', false)).to.eventually.be.rejected
        .then((error) => expect(error.message).to.equal('can\'t get cookie'));
    });
    it('execJiraQuery returns parsed response on 200 with CBA with expired cookie in the middle', () => {
      jira.config.authentication = 'cookie';
      let first = '{"resultset": "first"}';
      let second = '{"resultset": "second"}';
      let last = '{"resultset": "last"}';
      let expired = true;
      sandbox.stub(requestify, 'get', (url) => {
        if (url === '') {
          return Promise.resolve({code: 200, body: '{"startAt": 0,"total": 12,"maxResults": 5,"issues": [' + first + ']}'});
        } else if (url === '&startAt=5') {
          if ( expired === true ) {
            expired = false;
            return Promise.reject({code: 401, body: '{"message": "cookie expired!!"}'});
          } else {
            return Promise.resolve({code: 200, body: '{"startAt": 5,"total": 12,"maxResults": 5,"issues": [' + second + ']}'});
          }
        } else {
          return Promise.resolve({code: 200, body: '{"startAt": 10,"total": 12,"maxResults": 5,"issues": [' + last + ']}'});
        }
      });
      sandbox.stub(cookieAuth, 'login', () => {
        let header = {'set-cookie': ['mycookie=RUTN87766HG']};
        let body = '{"session": {"name": "mycookie"}}';
        return Promise.resolve({code: 200, headers: header, body: body});
      });
      return expect(jira.execJiraQuery('', true)).to.eventually.be.fulfilled
        .then((response) => {
          expect(response).to.have.property('startAt', 0);
          expect(response).to.have.property('total', 12);
          expect(response).to.have.property('maxResults', 12);
          expect(response).to.have.property('issues');
          expect(response.issues).to.be.an('array');
          expect(response.issues).to.have.length(3);
        });
    });
  });
});


test/test-cookieAuthentication.js
const test = require('unit.js');
const testThat = test.promise;
const sinon = require('sinon');
const chai = require('chai');
const chaiAsPromised = require('chai-as-promised');
const expect = chai.expect;
const rewire = require('rewire');
const cookie = rewire('../lib/cookieAuthentication.js');
const requestify = require('requestify');
const getBaseUrl = cookie.__get__('getBaseUrl');
const extract = cookie.__get__('extractSessionCookie');
let sandbox = null;

before(() => chai.use(chaiAsPromised));

beforeEach(() => sandbox = sinon.sandbox.create());

afterEach(() => sandbox.restore());


describe('getBaseUrl', () => {
  it('getBaseUrl() returns base url if valid url', () => testThat
    .given(() => 'http://www.dummyserver:23456/rest/api/2/issue/ID-5')
    .when((url) => getBaseUrl(url))
    .then((baseUrl) => expect(baseUrl).to.equal('http://www.dummyserver:23456/rest/'))
  );
  it('getBaseUrl() throw error if missing url', () => testThat
    .given()
    .when(() => () => getBaseUrl())
    .then((func) => expect(func).to.throw(Error))
  );
  it('getBaseUrl() throw error if invalid url', () => testThat
      .given(() => 'http://www.dummyserver:23456/bla/bla/bla')
      .when((url) => () => getBaseUrl(url))
      .then((func) => expect(func).to.throw(Error))
  );
});

describe('extractSessionCookie', () => {
  it('extractSessionCookie() throw error if cookie name not specified', () => testThat
    .given(() => 'fake="KHG8768"')
    .when((cookie) => () => extract(cookie))
    .then((func) => expect(func).to.throw(Error))
  );
  it('extractSessionCookie() throw error if cookies not specified', () => testThat
    .given()
    .when(() => () => extract())
    .then((func) => expect(func).to.throw(Error))
  );
  it('extractSessionCookie() throw error if cookies is not an array', () => testThat
    .given(() => ({cookies: 'fake="KHG987J"', name: 'fake'}))
    .then((data) => () => extract(data.cookies, data.name))
    .then((func) => expect(func).to.throw(Error))
  );
  it('extractSessionCookie() returns sessionCookie when only one cookie', () => testThat
    .given(() => ({cookies: ['cookie1="LKJHLKJ8768H"'], name: 'cookie1'}))
    .when((data) => extract(data.cookies, data.name))
    .then((cookie) => expect(cookie).to.equal('cookie1="LKJHLKJ8768H"'))
  );
  it('extractSessionCookie() throw error if cookie name not found in cookies', () => testThat
    .given(() => ({cookies: ['cookie1="LKJHLKJ8768H"'], name: 'cookie2'}))
    .when((data) => () => extract(data.cookies, data.name))
    .then((func) => expect(func).to.throw(Error))
  );
  it('extractSessionCookie() returns sessionCookie when multiple cookies', () => testThat
    .given(() => ({cookies: ['cookie1="LKJHLKJ8768H"', 'cookie2="KJHG76KHJB"', 'cookie3="JRS8MLKJKJF"', 'cookie4="JH8976HGFCJ"'], name: 'cookie3'}))
    .when((data) => extract(data.cookies, data.name))
    .then((cookie) => expect(cookie).to.equal('cookie3="JRS8MLKJKJF"'))
  );
  it('extractSessionCookie() returns sessionCookie when multiple cookies and empty cookie with cookie name ignored', () => testThat
    .given(() => ({cookies: ['cookie1="LKJHLKJ8768H"', 'cookie2="KJHG76KHJB"', 'cookie3=""', 'cookie3="JH8976HGFCJ"'], name: 'cookie3'}))
    .when((data) => extract(data.cookies, data.name))
    .then((cookie) => expect(cookie).to.equal('cookie3="JH8976HGFCJ"'))
  );
});

describe('getHeader', () => {
  it('getHeaders() builds header including cookie', () => testThat
    .given(() => 'cookiename=cookievalue')
    .when((sessionCookie) => cookie.getHeader(sessionCookie))
    .then((header) => {
      expect(header).to.have.property('cookie', 'cookiename=cookievalue');
      expect(header).to.have.property('Cache-Control', 'public, max-age=60');
    })
  );
});

describe('login', () => {
  it('login() rejects with if url not set', () => {
    return expect(cookie.login('user', 'password')).to.eventually.be.rejected
      .then((error) => {
        expect(error).to.be.an.instanceof(Error);
      });
  });
  it('login() rejects if no cookie is in header', () => {
    sandbox.stub(requestify, 'post', () => {
      let headers = {'set-cookie': ''};
      let response = '{"session": {"name": "studio.crowd.tokenkey"}}';
      return Promise.resolve({code: 200, headers: headers, body: response});
    });
    expect(cookie.login('dummy-user', 'dummy-password', 'http://www.dummyurl/rest/api/2/issue/ID-78')).to.eventually.be.rejected;
  });
  it('login() returns session Cookie', () => {
    sandbox.stub(requestify, 'post', () => {
      let headers = {'set-cookie': ['atlassian.xsrf.token=BGJJ-I70H-EYI8-6QPB|2ae8e3125acff97369f184a4530b59f9d983c12d|lout; Path=/; Secure', 'JSESSIONID=913F47DAFCA6D7FF09A65537D5BD3C5C; Path=/; Secure; HttpOnly', 'studio.crowd.tokenkey=""; Domain=.ulyssjira2.atlassian.net; Expires=Thu, 01-Jan-1970 00:00:10 GMT; Path=/; Secure; HttpOnly', 'studio.crowd.tokenkey=gW34EFQfK8Kbwpp6HkHmng00; Domain=.ulyssjira2.atlassian.net; Path=/; Secure; HttpOnly']};
      let response = '{"session": {"name": "studio.crowd.tokenkey"}, "loginInfo": {"failedLoginCount": 1, "loginCount": 230, "lastFailedLoginTime": "2017-01-17T10:20:43.467+0100", "previousLoginTime": "2017-01-17T17:11:46.798+0100"}}';
      return Promise.resolve({code: 200, headers: headers, body: response});
    });
    return expect(cookie.login('dummy-user', 'dummy-password', 'http://www.dummyurl/rest/api/2/issue/ID-78')).to.eventually.be.fulfilled
      .then((cookie) => {
        expect(cookie).to.equal('studio.crowd.tokenkey=gW34EFQfK8Kbwpp6HkHmng00');
      });
  });
});


lib/cookieAuthentication.js
const requestify = require('requestify');

let getBaseUrl = (url) => {
  if (!url) {
    throw new Error('url not specified - can\'t get base url');
  }
  let index = url.indexOf('rest/');
  if (index === -1) {
    throw new Error('Invalid url (can\'t find \'rest/\' string) - can\'t get base url');
  }
  let baseUrl = url.substr(0, index + 5);

  return baseUrl;
};

let extractSessionCookie = (cookies, cookieName) => {
  if (!cookieName) {
    throw new Error('cookie name not specified - can\'t extract session cookie');
  }
  if (!cookies) {
    throw new Error('cookies not specified - can\'t extract session cookie');
  }
  if (!Array.isArray(cookies)) {
    throw new Error('cookies is not an array - can\'t extract session cookie');
  }

  try {
    var sessionCookie = cookies.find((cookie) => { // eslint-disable-line no-var
      let name = cookie.split('=')[0];
      let value = cookie.split('=')[1].split(';')[0];
      return ((name === cookieName) && (value != '""'));
    }).split(';')[0];
  } catch (error) {
    throw new Error('cookie name not found in cookies - can\'t extract session cookie');
  }

  return sessionCookie;
};

module.exports = {
  /**
   * Login to Jira is order to retrieve the session cookie
   * @param  {String} userName the user name to use to log in Jira
   * @param  {String} password the password of the user used to log in Jira
   * @param  {String} url      the url of the request to execute - it used to compute the login url
   * @return {Promise}         Returns the sessions cookie as a {String} (<cookie name>=<cookie value>)
   */
  login: (userName, password, url) => new Promise((resolve, reject) => {
    try {
      let loginUrl = getBaseUrl(url) + 'auth/1/session';
      let body = {'username': userName, 'password': password};
      let options = {
        cache: {
          cache: false,
        },
      };

      requestify.post(loginUrl, body, options)
        .then((response) => {
          resolve(extractSessionCookie(response.headers['set-cookie'], JSON.parse(response.body).session.name));
        })
        .catch((error) => {
          reject(new Error('Can\'t get cookie [' + error.message + ']'));
        });
    } catch (error) {
      reject(new Error('Can\'t get cookie [' + error.message + ']'));
    }
  }),
  /**
   * Builds the header to be added to requests using cookie based authentication
   * @param  {String} sessionCookie the cookie: '<cookie name>=<cookie value>'
   * @return {Object}               the header object
   */
  getHeader: (sessionCookie) => {
    return {'cookie': sessionCookie, 'Cache-Control': 'public, max-age=60'};
  },
};


lib/jira.js
const requestify = require('requestify');
const config = require('./config');
const cookieAuthentication = require('./cookieAuthentication');
let sessionCookie = null;

/**
 * Used to build an url to get more results than 50 when required
 * @param  {String} oldUrl  the previous executed url
 * @param  {Number} startAt new start for the url to build
 * @return {String}         the new url to get next results
 */
let buildUrlToGetNextPage = (oldUrl, startAt) => {
  let url = '';

  if (oldUrl.search('startAt') === -1) {
    url = oldUrl + '&startAt=' + startAt;
  } else {
    let index = oldUrl.search('&startAt=');
    url = oldUrl.replace('&startAt=' + oldUrl.substr(index + 9, oldUrl.length - index + 9), '&startAt=' + startAt);
  }

  return url;
};
/**
 * Builds the header that will be added all requests to Jira
 * @param  {String} url the url that will be executed - this is mandatory for cookie based authentication and useless for basic authentication
 * @return {Promise}     returns the header object and rejects as an {Error}
 */
let getHeader = (url) => new Promise((resolve, reject) => {
  let user = process.env.JIRA_USER;
  let password = process.env.JIRA_PASSWORD;

  if (module.exports.config.authentication === 'cookie') {
    if (sessionCookie) {
      resolve(cookieAuthentication.getHeader(sessionCookie));
    } else {
      cookieAuthentication.login(user, password, url)
        .then((cookie) => {
          sessionCookie = cookie;
          resolve(cookieAuthentication.getHeader(sessionCookie));
        })
        .catch((error) => {
          reject(error);
        });
    }
  } else {
    let token = 'Basic ' + new Buffer(user + ':' + password).toString('base64');
    resolve({
      'Authorization': token,
      'Cache-Control': 'public, max-age=60',
    });
  }
});

let execJiraQueryUntilRecursive = (url, retrieveAllPages, resolve, reject, maxRetry, count) => {
  console.log('try to execute request (' + url + '): ' + count + '/' + maxRetry);

  if (count === maxRetry) {
    console.error('too much retries, stopping here');
    reject({code: -1, url: url, message: 'Too much retries'});
  } else {
    getHeader(url)
      .then((header) => {
        let options = {
          cache: {
            cache: false,
            expires: 30000,
          },
          headers: header,
        };

        requestify.get(url, options)
          .then((response) => {
            console.log('Request OK, code = ' + response.code);
            let jsonBody;
            try {
              jsonBody = JSON.parse(response.body);
            } catch (error) {
              console.error('Error parsing response. Url = ' + url + ', Error = ' + error.message);
              return Promise.reject({code: -1, url: url, message: error.message});
            }
            return jsonBody;
          })
          .then((response) => {
            if (retrieveAllPages && response.startAt !== null && response.startAt + response.maxResults <= response.total) {
              let startAt = response.startAt + response.maxResults;
              let nextPageUrl = buildUrlToGetNextPage(url, startAt);
              execJiraQueryUntil(nextPageUrl, true, 5)  // eslint-disable-line no-use-before-define
                .then((nextPageResponse) => {
                  response.issues = response.issues.concat(nextPageResponse.issues);
                  response.maxResults = response.total;
                  response.startAt = 0;
                  resolve(response);
                })
                .catch((error) => reject(error));
            } else {
              resolve(response);
            }
          })
          .catch((error) => {
            if (error.code === 401 && module.exports.config.authentication === 'cookie') { // cookie expired, gets a new one
              console.log('Error executing request, cookie expired, but I\'ll log in again');
              sessionCookie = null;
              setTimeout(execJiraQueryUntilRecursive, module.exports.config.retryTimeout, url, retrieveAllPages, resolve, reject, maxRetry, count + 1);
            } else if ((error.code === 'ECONNRESET') || (error.code === 'ENOTCONN') || (error.body === 'timeout exceeded') || (Math.floor(error.code / 100) === 5)) {
              console.error('Error executing request, but I\'ll try again... Url = ' + url + ', Error = ' + JSON.stringify(error, null, '\t'));
              setTimeout(execJiraQueryUntilRecursive, module.exports.config.retryTimeout, url, retrieveAllPages, resolve, reject, maxRetry, count + 1);
            } else {
              reject({code: error.code, url: url, message: error.body ? error.body : error.message});
            }
          });
      })
      .catch((error) => {
        reject({code: -1, url: url, message: error.message});
      });
  }
};

let execJiraQueryUntil = (url, retrieveAllPages, maxRetry) => {
  let count = 1;
  return new Promise((resolve, reject) => {
    execJiraQueryUntilRecursive(url, retrieveAllPages, resolve, reject, maxRetry, count);
  });
};

module.exports = {
  /**
   * Configuration
   */
  config,
  /**
   * Return true if Jira credentials are not set as environment variables (JIRA_USER and JIRA_PASSWORD)
   * @return {boolean} true if credentials are missing
   */
  areJiraCredentialsMissing: () => {
    let user = process.env.JIRA_USER;
    let password = process.env.JIRA_PASSWORD;
    let check = (!user || !password);

    return check;
  },
  /** Execute a Jira Rest api and returns the response as a Promise.
   * When the request returns an error 5xx, it is retried 5 times with a 1s timeout in between before returning
   * @param  {String} url               the url to execute
   * @param  {boolean} retrieveAllPages if true and if the query returns more than 50 items, all other items are automatically retrieved and
   * @return {Promise}                  returns the json response, or in case of error an object with 3 properties: code {Number}, url {String}, message {String}
   */
  execJiraQuery: (url, retrieveAllPages) => new Promise((resolve, reject) => {
    execJiraQueryUntil(url, retrieveAllPages, config.maxRetry)
      .then((response) => resolve(response))
      .catch((error) => reject(error));
  }),
};
